Dev: migration: implement corosync.conf migration for corosync 3 (jsc#PED-8252)

nicholasyang2022 opened the pull request:
This pull request implements migrating `corosync.conf` from corosync 2 to corosync 3, including:

* migrate deprecated transport `udpu` and `udp` to `knet`
* replace removed feature _RRP_ with _knet multilink_
* if `crypto_hash` is set to `sha1`, which is the default value in corosync 2, upgrades it to the new default in corosync 3, `sha256`.

---

# Use Cases

## Check When All Green for Migration

```
> sudo crm cluster health sles16
------ node: localhost ------
[INFO] Checking dependency version...
[INFO] Checking used corosync features...
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use knet.

------ cib ------
[WARN] The CIB is not validated with the latest schema version.
       * Latest version:  3.10
       * Current version: 3.9
[INFO] Checking used resource agents...

------ node: ha-2-2 ------
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use knet.

****** summary ******
[INFO] Please run "crm cluster health sles16 --fix" on on any one of above nodes.
[PASS] This cluster is good to migrate to SLES 16.

```

## Check When Already Migrated

```
> sudo crm cluster health sles16
------ node: localhost ------
[INFO] Checking dependency version...
[INFO] Checking used corosync features...

------ cib ------
[INFO] Checking used resource agents...

------ node: ha-2-2 ------

****** summary ******
[INFO] This cluster works on SLES 16. No migration is needed.
```

## Run Migration

```
> sudo crm cluster health sles16 --fix
suse@ha-2-1:~> sudo crm cluster health sles16 --fix
------ node: localhost ------
[INFO] Checking dependency version...
[INFO] Checking used corosync features...
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use knet.

------ cib ------
[WARN] The CIB is not validated with the latest schema version.
       * Latest version:  3.10
       * Current version: 3.9
[INFO] Checking used resource agents...

------ node: ha-2-2 ------
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use knet.

INFO: Starting migration...
INFO: Migrating corosync configuration...
INFO: Upgrade totem.transport to knet.
INFO: Upgrade totem.crypto_hash from "sha1" to "sha256".
INFO: Finish migrating corosync configuration. The original configuration is renamed to corosync.conf.bak
INFO: Finished migration.
```

## Try to Run Migration When Already Migrated

```
suse@ha-2-1:~> sudo crm cluster health sles16 --fix
------ node: localhost ------
[INFO] Checking dependency version...
[INFO] Checking used corosync features...

------ cib ------
[WARN] The CIB is not validated with the latest schema version.
       * Latest version:  3.10
       * Current version: 3.9
[INFO] Checking used resource agents...

------ node: ha-2-2 ------

INFO: This cluster works on SLES 16 with some warnings. Please fix the remaining warnings manually.
```

liangxin1300 commented:
In 
```
# crm cluster health sles16
------ localhost ------
[INFO] Checking dependency version...
[FAIL] Pacemaker version not supported
       Supported version: 3 <= Pacemaker
       Actual version:    Pacemaker == 2.1.9
```
`Supported version: 3 <= Pacemaker` better changed similar as `Supported version: Pacemaker >= 3`

liangxin1300 commented:
```
# crm cluster health sles16
------ localhost ------
[INFO] Checking dependency version...
[FAIL] Pacemaker version not supported
       Supported version: 3 <= Pacemaker
       Actual version:    Pacemaker == 2.1.9
[INFO] Checking service status...
[FAIL] Cluster services are running
       * corosync
       * pacemaker
[INFO] Checking used corosync features...

------ alp-3 ------



------ alp-2 ------
```
No output for remote nodes

zzhou1 commented:
> [FAIL] Cluster services are running
>        * corosync
>        * pacemaker

FAIL as the fatal error sounds too strong to me. 

No matter it is corosync2 or corosync3 environment, nothing wrong to run `crm cluster health sles16` no matter the cluster is running or not.

liangxin1300 commented:
## Format discussion
#### This is pre-migration on SLE15SP6
```
# crm cluster health sles16 
------ corosync @ 15sp6-2 ------
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use "knet".

------ corosync @ localhost ------
[INFO] Checking dependency version...
[INFO] Checking service status...
[INFO] Checking used corosync features...
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use "knet".
------ cib ------
[INFO] Checking used resource agents...

****** summary ******
```
#### This is migration on Opensuse:TW
```
# crm cluster health sles16 
------ localhost ------
[INFO] Checking dependency version...
[INFO] Checking service status...
[INFO] Checking used corosync features...

------ cib ------
[INFO] Checking used resource agents...

------ alp-2 ------

****** summary ******
```

## Suggestions
- Keep the format `<component> @ <node>` for each check scope header, on both pre-migration and migration codes
- Replace `cib` with the same format, or user might think `cib` is a node name

nicholasyang2022 commented:
> * Keep the format <component> @ <node> for each check scope header

Then you will get:

```
------ corosync @ ha-2-2 ------
[FAIL] Cluster services are running
       * corosync
       * pacemaker
```

liangxin1300 commented on crmsh/cibquery.py:
```diff
@@ -0,0 +1,26 @@
+    )
+
+
+def has_primitive_filesystem_with_fstype(cib: lxml.etree.Element, fstype: str) -> bool:
```
For commonly use purpose, I do think this function can be renamed more commonly, like `has_primitive_ra_with`, with parameters like type and nvpair's name and value

nicholasyang2022 replied:
I do know xml queries can be more flexible, just like JAXB in java. But a complete, user-friendly and reliable implementation needs too much work and is impossible to be included in crmsh. I incline to implement some routines only when we really need to call it from many places. Do you have some real use cases?

liangxin1300 commented on crmsh/migration.py:
```diff
@@ -0,0 +1,655 @@
+    """Return:
+    * 0 if the cluster is already on SLES 16.
+    * 1 if the cluster is good to migrate to SLES 16.
+    * 2 if the cluster is not good to migrate to SLES 16."""
```
Better to create enum to make these return codes more describable, since other function like `migration` will also use these values

liangxin1300 commented on crmsh/migration.py:
```diff
@@ -0,0 +1,655 @@
+        config = corosync_config_format.DomParser(f).dom()
+    logger.info('Migrating corosync configuration...')
+    migrate_corosync_conf_impl(config)
+    shutil.copy(conf_path, conf_path + '.bak')
```
crmsh leave corosync.conf.bak here, do not remove it after migration, or do not mention that this bak file has been created?

nicholasyang2022 commented:
Recent Changes:

* Do not check whether cluster services are running.
* Changed the section header `--- hanode1 ---` to `--- node: hanode1 ---`
* Changed the message `Supported version: 3 <= Corosync` to `Supported version: Corosync >= 3`
* Added a message about corosync.conf backup.
* Added a message about problems need manual fix: `INFO: This cluster works on SLES 16 with some warnings. Please fix the remaining warnings manually.`

liangxin1300 commented on crmsh/migration.py:
```diff
@@ -0,0 +1,673 @@
+        interface.pop('ttl', None)
+        ringnumber = interface.pop('ringnumber', None)
+        if ringnumber is not None:
+            interface['linknumber'] = ringnumber
```
Redundant codes with `migrate_udpu`, can be improved as more consolidate

liangxin1300 reviewed (approved):
Great work
Thanks!
